rfc/RFC-0001-dynamic-embedding-with-PS.md

```diff
@@ -0,0 +1,95 @@
+
+1. The remote PS and the embedding of TorchRec clusters form a hot-cold separation of the features naturally.
+2. There are mature industrial solutions for CPU PS that could support nearly unlimited storage and quickly send the training model weight to inference.
+3. PS could decouple the training and inference process, making it possible to migrate to TorchRec incrementally.
+4. The cost of CPU PS is normally much lower than GPU servers.
+
+**NOTE**: The PS here is only responsible for storing the embedding features, similar to KV storage like redis. All the model training computations (forward, backward, and optimizer) are done within the TorchRec GPU cluster.
+
+Under such a setting, the distributed embedding in GPU clusters would serve as a cache of the whole system. For each feature, it will have a `global_id`, representing its id in PS, and if it is currently stored in GPU embedding, a `cache_id`, representing its id in the GPU embedding. When a new data sample arrives, it needs to convert the `global_id` to `cache_id` and update the “cache” if necessary. The following is pseudocode of the preprocessing method:
+
+```python
+def preprocess(global_id):
+    if global_id in cache_index:
+        cache_id = cache_index[global_id]
+    else:
+        if cache_index.full():
+           # push the old features back to PS
+           evict(cache_index, embedding)
+        cache_id = cache_index.get_available_cache_id()
+        # fetch the feature from PS and insert to the
+        # corresponding position in embedding
+        insert(cache_index, embedding, cache_id)
+    return cache_id
+```
+
+**NOTE**: The above code snippet is only for showing the possible procedures involved, the real implementation should work in a more performant way.
+
+To implement this architecture, we need to clarify some implementation details:
+
+1. How to move the features fetched from PS into GPU embedding?
+
+    Because TorchRec supports multiple sharding schema, different segments of an embedding feature may be stored on different devices, when given a `cache_id`, we need to cut it into segments and move those segments to their actual locations in the ShardedTensor.
+    We may store features in PS according to the sharding schema, i.e. save segments of the feature separately, so that each device can only fetch the segment that will be in its shard.
+
+2. How to have a consistent cache index across devices?
+
+    During training, the same `cache_id` on different devices must be storing the feature with the same `global_id`. To ensure strong consistency, we could use let each device only maintain the cache info of the features that are stored locally, and communicate with other devices if cache info of other features is needed.
```

Comment:
I'm assuming this means locks/mutexes

Reply:
In general, yes. The mapping between `cache_id` and `global_id` should be locked.

I think it could be a lock-free implementation, since generally it just an `std::vector<atomic<long>> cache_ids_to_global_ids`. 

Instead of synchronizing the map, we think to synchronize the async job is simpler. Each data batch feed to `TorchRec` just like 

```
struct DataSampleToTorchRec {
  std::future<void()> fetch_job_complete_;
  DataBacth transformed_data_batch_;
};
```